atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/CassandraGetCandidateCellsForSweepingImpl.java

```diff
@@ -0,0 +1,155 @@
+import com.google.common.collect.Lists;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweepingRequest;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.ImmutableCandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServices;
+import com.palantir.atlasdb.keyvalue.cassandra.paging.CassandraRawCellValue;
+import com.palantir.atlasdb.keyvalue.cassandra.paging.CellPager;
+import com.palantir.common.annotation.Output;
+import com.palantir.util.Pair;
+
+import gnu.trove.list.TLongList;
+import gnu.trove.list.array.TLongArrayList;
+
+/*
+ * Simply use the CellPager to iterate over raw cells in a table and group the returned entries by the cell key.
+ */
+public class CassandraGetCandidateCellsForSweepingImpl {
+    private final CellPager cellPager;
+
+    public CassandraGetCandidateCellsForSweepingImpl(CellPager cellPager) {
+        this.cellPager = cellPager;
+    }
+
+    public Iterator<List<CandidateCellForSweeping>> getCandidateCellsForSweeping(
+            TableReference tableRef,
+            CandidateCellForSweepingRequest request,
+            ConsistencyLevel consistencyLevel) {
+        Iterator<List<CassandraRawCellValue>> rawIter = cellPager.createCellIterator(
+                tableRef,
+                request.startRowInclusive(),
+                request.batchSizeHint().orElse(AtlasDbConstants.DEFAULT_SWEEP_CANDIDATE_BATCH_HINT),
+                consistencyLevel);
+        return new CellGroupingIterator(rawIter, request);
+    }
+
+    private static class CellGroupingIterator extends AbstractIterator<List<CandidateCellForSweeping>> {
+        private final Iterator<List<CassandraRawCellValue>> rawIter;
+        private final CandidateCellForSweepingRequest request;
+        private Cell currentCell = null;
+        private final TLongList currentTimestamps = new TLongArrayList();
+        private boolean currentLatestValEmpty;
+        private long numCellTsPairsExamined = 0;
+        private boolean end = false;
+
+        CellGroupingIterator(Iterator<List<CassandraRawCellValue>> rawIter, CandidateCellForSweepingRequest request) {
+            this.rawIter = rawIter;
+            this.request = request;
+        }
+
+        @Override
+        protected List<CandidateCellForSweeping> computeNext() {
+            if (end) {
+                return endOfData();
+            } else {
+                List<CandidateCellForSweeping> candidates = Lists.newArrayList();
+                while (candidates.isEmpty() && rawIter.hasNext()) {
+                    List<CassandraRawCellValue> cols = rawIter.next();
+                    for (CassandraRawCellValue col : cols) {
+                        processColumn(col, candidates);
```

Comment:
I now se that `collectCandidatesForSweeping` returns 0 or 1 candidates - so we could have that method return an `Optional` instead of a `List`, and maybe the method would be better named something like `createCandidateForSweepingIfNecessary`